### PR TITLE
Config

### PR DESCRIPTION
**Pull request by wolfeman67**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -43,7 +44,8 @@ function clientLoaded (client) {
  * @param {Object} err - error from application.
  */
 function errHandler(err) {
-  if(err.name === 'EarlyHangup' || err.name === 'HangupFailure') {
+  if(err.name === 'EarlyHangup' || err.name === 'HangupFailure' ||
```
Comment (samuelg) [inline app.js]: That is already a bit unwieldy and will only grow, I would move that check to a function. fatalError() perhaps?

File: app.js [hunk 2]
```diff
@@ -29,8 +29,9 @@ function clientLoaded (client) {
   client.start('sla');
   client.on('StasisStart', function(event, channel) {
     if(!isDialed(event.args[0])) {
-      var bridgeName = event.args[0];
-      sla(client, channel, bridgeName)
+      var extension = event.args[0];
+      var confFile = process.argv[2];
```
Comment (samuelg) [inline app.js]: What happens if the conf file is not provided by the CLI?

File: res/config.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
+{
+  "sharedExtensions": [
+  {
+    "42": {
+      "trunks": [
+        {"42":
```
Comment (samuelg) [inline res/config.json]: This is hard to read. I would look at prettifying the json file so it's in a common format.

File: tests/testConfigs/multipleEndpoints.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{                                                                                
+  "sharedExtensions": [                                                          
+  {                                                                              
+    "42": {                                                             
+      "trunks": [                                                                
+        {"42":                                                                   
+          {                                                                      
+            "stations": [                                                        
+            {"endpoint": "SIP/phone1"
```
Comment (samuelg) [inline tests/testConfigs/multipleEndpoints.json]: Look at formatting here as well.

File: tests/testConfigs/noEndpoints.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
+{                                                                                
+  "sharedExtensions": [                                                          
+  {                                                                              
+    "42": {                                                             
+      "trunks": [                                                                
+        {"42":
```
Comment (samuelg) [inline tests/testConfigs/noEndpoints.json]: Look at formatting here as well.

File: tests/testConfigs/singleEndpoint.json [hunk 6]
```diff
@@ -0,0 +1,17 @@
+{                                                                                
+  "sharedExtensions": [                                                          
+  {                                                                              
+    "42": {                                                             
+      "trunks": [                                                                
+        {"42":
```
Comment (samuelg) [inline tests/testConfigs/singleEndpoint.json]: Look at formatting here as well.

File: lib/dal.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+ * @param {string} confFile - the path and filename to the configuration file
+ * @param {string} name - the name of the sharedExtension to access
+ * @return {Q} - Q promise object
+ */
+
+var getSharedExtension = function(confFile, name) {
+  var defer = Q.defer();
+  var sharedExtension = {};
+  sharedExtension.pos = -1;
+  return getConfig(confFile).then(function(data) {
```
Comment (samuelg) [inline lib/dal.js]: You are putting an implementation detail directly in the API you expose. If you need to add database support later, you'll have a lot more work to separate this out from the functionality of returning a shared extension object. Also, what if you wanted to support both config files and a database?

You should have an internal only API for getting the data that can be modified to support various back-ends instead of having this inside the API you expose.

File: lib/dal.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+/** Attempts to access configuration file data for the shared extension
+ * @param {string} confFile - the path and filename to the configuration file
+ * @param {string} name - the name of the sharedExtension to access
+ * @return {Q} - Q promise object
+ */
+
+var getSharedExtension = function(confFile, name) {
+  var defer = Q.defer();
+  var sharedExtension = {};
+  sharedExtension.pos = -1;
```
Comment (samuelg) [inline lib/dal.js]: You could also have a boolean called found to determine this since -1 is rather cryptic.

File: lib/dal.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+  this.trunks.forEach(function(element, i) {
+    if(element[trunkName]) {
+      trunk.getAllStations = getAllStations;
+      trunk.trunkName = trunkName;
+      trunk.stations = element[trunk.name].stations;
+      trunk.pos = i;
+      defer.resolve(trunk);
+    }
+  });
+  if(!trunk.pos) {
```
Comment (samuelg) [inline lib/dal.js]: I would use a boolean here.

File: lib/sla.js [hunk 10]
```diff
@@ -98,46 +99,102 @@ function addChannelsToBridge(inbound, dialed, bridge) {
...
+    toKill.forEach(function (unanswered) {
+      unanswered.removeListener('ChannelDestroyed', dialedChannelHungup);
+      var hangup = Q.denodeify(unanswered.hangup.bind(unanswered));
+      hangup();
+    });
+    defer.resolve(line);
+  }
+
+  function dialedChannelHungup(event) {
+    numToHangup -= 1;
```
Comment (samuelg) [inline lib/sla.js]: Why not remove the dialed channel from the dialed array and checking the length of the array instead? You could also have a dialed object with an array of channels that get updated when hungup and a areAllHungup() method.

File: lib/sla.js [hunk 11]
```diff
@@ -98,46 +99,102 @@ function addChannelsToBridge(inbound, dialed, bridge) {
...
+  }
+
+  trunkData.getAllStations()
+    .then(function (stations) {
+      if (stations.length) {
+        stations.forEach(function() {
+          dialed.push(client.Channel());
+        });
+        numToHangup = dialed.length;
+        dialed.forEach(function (channel, i, dialed) {
```
Comment (samuelg) [inline lib/sla.js]: i is not a very descriptive variable name. At the very least use index.

File: lib/sla.js [hunk 12]
```diff
@@ -98,46 +99,102 @@ function addChannelsToBridge(inbound, dialed, bridge) {
...
+          originate.push(Q.denodeify(channel.originate.bind(dialed[i])));
+          originate[i]({endpoint: stations[i].endpoint, app: 'sla',
```
Comment (samuelg) [inline lib/sla.js]: Why keep the bound originate methods around?

File: lib/sla.js [hunk 13]
```diff
@@ -98,46 +99,102 @@ function addChannelsToBridge(inbound, dialed, bridge) {
...
+    });
     defer.resolve('Inbound channel hungup');
-  });
+  }
+
+  trunkData.getAllStations()
+    .then(function (stations) {
+      if (stations.length) {
+        stations.forEach(function() {
+          dialed.push(client.Channel());
```
Comment (samuelg) [inline lib/sla.js]: If you create an object with a station and a dialed channel, you'll be able to just iterate over that and not use an index anymore.

File: lib/sla.js [hunk 14]
```diff
@@ -98,46 +99,102 @@ function addChannelsToBridge(inbound, dialed, bridge) {
...
+        });
+        numToHangup = dialed.length;
+        dialed.forEach(function (channel, i, dialed) {
+          originate.push(Q.denodeify(channel.originate.bind(dialed[i])));
+          originate[i]({endpoint: stations[i].endpoint, app: 'sla',
+            appArgs: 'dialed', timeout: 10})
```
Comment (samuelg) [inline lib/sla.js]: This should be indented 2 more spaces.

File: lib/sla.js [hunk 15]
```diff
@@ -98,46 +99,102 @@ function addChannelsToBridge(inbound, dialed, bridge) {
...
-function checkIfNum (string) {
-  if(!isNaN(parseInt(string))) {
-    return true;
-  }
-  else {
-    return false;
-  }
+function getTrunkData (configurationFile, extension) {
+  var defer = Q.defer();
+  dal.getSharedExtension(configurationFile, extension)
```
Comment (samuelg) [inline lib/sla.js]: There's no need for a deferred here. Just return dal.getSharedExtension and within it either return a value on success or throw an error if not.

Deferred should only be created explicitly if you are creating a promise supported API or if you want the result of an event listener firing to resolve a promise. Since dal.getSharedExtension is already a promise supported API there's no reason to create a deferred yourself.

File: res/config.json [hunk 16]
```diff
@@ -0,0 +1,22 @@
+{
+  "sharedExtensions": [
+    {
+      "42": {
+        "trunks": [
+          {
+            "42": {
+              "stations": [
```
Comment (matt-jordan) [inline res/config.json:8]: While it may be better to deal with this when you start manipulating outbound dial, stations should not be conceptually a child of trunks. Trunks are a type of endpoint that participate in a shared extension; they are a separate concept from the stations that they serve.

File: lib/dal.js [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+/** Attempts to access configuration file data for the shared extension
+ * @param {string} confFile - the path and filename to the configuration file
+ * @param {string} name - the name of the sharedExtension to access
+ * @return {Q} - Q promise object
+ */
+
+var getSharedExtension = function(confFile, name) {
+  var defer = Q.defer();
+  var sharedExtension = {};
```
Comment (matt-jordan) [inline lib/dal.js:14]: From a stylistic perspective, it is often nice to have a line break between your variable declarations and the code that acts on them.

``` javascript
var defer = Q.defer();
var sharedExtension = {};

sharedExtension.found = false;
...
```

File: lib/dal.js [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+  sharedExtension.found = false;
+  return fetchJSON(confFile).then(function(data) {
+    data.sharedExtensions.forEach(function(extension, i) {
+      if(extension[name]) {
```
Comment (matt-jordan) [inline lib/dal.js]: Space between if and conditional:

``` javascript
if (extension[name]) {
```

File: app.js [hunk 19]
```diff
@@ -38,12 +42,25 @@ function clientLoaded (client) {
   });
 }
 
+/** Utility function for seeing if an error is fatal (crashes the program)
+ * @param {Object} err - the error in question.
+ * @return {boolean} - if the error is fatal or not
+ */
+function isFatal(err) {
+  if(err.name === 'EarlyHangup' || err.anme === 'HangupFailure' ||
```
Comment (matt-jordan) [inline app.js]: Space between if and conditional:

``` javascript
if (err.name ...
```

File: app.js [hunk 20]
```diff
@@ -29,8 +33,8 @@ function clientLoaded (client) {
   client.start('sla');
   client.on('StasisStart', function(event, channel) {
     if(!isDialed(event.args[0])) {
-      var bridgeName = event.args[0];
-      sla(client, channel, bridgeName)
+      var extension = event.args[0];
+      sla(client, channel, confFile, extension)
```
Comment (matt-jordan) [inline app.js]: Your parameter ordering here is a bit odd:

1: The ARI client (okay)
2: The channel entering the shared extension (passed in from Stasis, okay)
3: The config file for the shared extension (but what shared extension?)
4: The actual shared extension (passed in from Stasis...)

From an ease of use perspective, you should order your parameters such that they flow logically. One ordering would be:

``` javascript
client, confFile, channel, extension
```

In other words: lead with the things that are unlikely to ever change (client, confFile) and follow with the more mutable pieces of data.

File: app.js [hunk 21]
```diff
@@ -6,11 +6,15 @@ var sla = require('./lib/sla.js');
...
-  .catch(errHandler)
-  .done();
+var confFile;
+if (confFile = process.argv[2]) {
+  connect('http://127.0.0.1:8088', 'user', 'pass')
+    .then(clientLoaded)
+    .catch(errHandler)
+    .done();
+} else {
+  console.error('No configuration file specified');
```
Comment (matt-jordan) [inline app.js]: Is it fatal to not provide or find a configuration file? I would imagine this should actually bail out of the application.

File: lib/sla.js [hunk 22]
```diff
@@ -98,65 +99,129 @@ function addChannelsToBridge(inbound, dialed, bridge) {
...
-        return unanswered.id !== line.id;
-      });
-      toKill.forEach(function (unanswered) {
-        var hangup = Q.denodeify(unanswered.hangup.bind(unanswered));
-        hangup();
-      });
-      defer.resolve(line);
+  var dialed = [];
+
+  function dialedEnteredStasis(event, line) {
```
Comment (matt-jordan) [inline lib/sla.js:116]: Document your function

File: lib/sla.js [hunk 23]
```diff
@@ -98,65 +99,129 @@ function addChannelsToBridge(inbound, dialed, bridge) {
...
-    dialed[index].once('ChannelDestroyed', function(event) {
-      originate.pop(dialed[index]);
-      if(originate.length === 0) {
-        defer.resolve('All dialed channels hungup');
-        var hangup = Q.denodeify(inbound.hangup.bind(inbound));
-        hangup();
+    defer.resolve(line);
+  }
+
+  var allAreHungup = function(dialed) {
```
Comment (matt-jordan) [inline lib/sla.js:137]: Document this function

File: lib/sla.js [hunk 24]
```diff
@@ -98,65 +99,129 @@ function addChannelsToBridge(inbound, dialed, bridge) {
...
+   if (dialed.length) {
+     return true;
+   } else {
+     return false;
+   }
+  };
+
+  var findInDialed = function(channel) {
```
Comment (matt-jordan) [inline lib/sla.js:147]: Document this function

File: lib/sla.js [hunk 25]
```diff
@@ -98,65 +99,129 @@ function addChannelsToBridge(inbound, dialed, bridge) {
...
+    var position = -1;
+    dialed.forEach(function(object, index) {
+      if(channel === dialed.channel) {
+        position = index;
+      }
+    });
+    return position;
+  };
+
+  function dialedChannelHungup(event) {
```
Comment (matt-jordan) [inline lib/sla.js:162]: Document this function

File: lib/sla.js [hunk 26]
```diff
@@ -98,65 +99,129 @@ function addChannelsToBridge(inbound, dialed, bridge) {
...
+            'dialed.  Element not found.'));
+    }
+        
+    if(allAreHungup(dialed)) {
+      inbound.removeListener('ChannelHangupRequest', inboundHungup);
+      defer.resolve('Dialed channels hungup');
+    }
+  }
+
+  function inboundHungup(event, line) {
```
Comment (matt-jordan) [inline lib/sla.js:184]: Document this function

File: lib/fetchJSON.js [hunk 27]
```diff
@@ -0,0 +1,15 @@
+var fs = require('fs');
+var Q = require('q');
+
+/** Reads from the configuration file and returns data
```
Comment (matt-jordan) [inline lib/fetchJSON.js]: From a stylistic perspective, the description should go on the next line:

``` javascript
/**
 * Read from the configuration file and returns data
```

This finding applies to multiple comment blocks

File: res/config.json (hunk 16 above)
Comment (wolfeman67) [inline res/config.json:8]: I have an idea of how to fix this, but I'll fix it in the next issue.

File: app.js [hunk 28]
```diff
@@ -38,12 +42,25 @@ function clientLoaded (client) {
...
+      }
+}
+
 /**
  * Handles errors found in application.
  * @param {Object} err - error from application.
  */
 function errHandler(err) {
-  if(err.name === 'EarlyHangup' || err.name === 'HangupFailure') {
+  if (isFatal(err)) {
```
Comment (samuelg) [inline app.js]: It looks like you only log the error if it is fatal, but blow up if it is not fatal.

File: app.js [hunk 29]
```diff
@@ -38,12 +42,25 @@ function clientLoaded (client) {
   });
 }
 
+/** Utility function for seeing if an error is fatal (crashes the program)
+ * @param {Object} err - the error in question.
+ * @return {boolean} - if the error is fatal or not
+ */
+function isFatal(err) {
+  if (err.name === 'EarlyHangup' || err.anme === 'HangupFailure' ||
```
Comment (samuelg) [inline app.js]: Since it evaluates to a boolean already, you can just use:

return err.name === 'EarlyHangup' || err.anme === 'HangupFailure' || ....;

File: lib/sla.js [hunk 30]
```diff
@@ -98,75 +99,156 @@ function addChannelsToBridge(inbound, dialed, bridge) {
...
+    defer.resolve(line);
+  }
+  /**
+   * Utility function for cheking if all dialed channels have been hungup
+   * @param {Array} dialed - the array of channels; the length is checked
+   * @return {boolean} - whether or not the dialed array is empty
+   */
+
+  var allAreHungup = function(dialed) {
+   if (dialed.length) {
```
Comment (samuelg) [inline lib/sla.js]: You can just use:

return dialed.length;

to get the same result. Although I don't think it's doing what you think it's doing. Any integer above 0 will evaluate to true, so if the array has channels that have not been hungup yet, it will evaluate that all channels have been hungup. You should do !dialed.length instead so it only evaluates to true once the array has had all its elements removed.

File: lib/sla.js [hunk 31]
```diff
@@ -98,75 +99,156 @@ function addChannelsToBridge(inbound, dialed, bridge) {
...
+      return Q.resolve(trunk);
+    })
+    .catch(function(invalid) {
+      // This will only be called if trunk and extension are not the same
+      // Shouldn't be right now
+      return Q.reject(new CustomError('InvalidTrunkSpecification', invalid));
+    });
+  })
+  .catch(function(invalid) {
+    if (invalid === 'Invalid specified extension: ' + extension) {
```
Comment (samuelg) [inline lib/sla.js]: You should check invalid.name for the Exception reason.

File: lib/sla.js [hunk 32]
```diff
@@ -183,37 +265,37 @@ function CustomError(name, message) {
...
+              }
+              else {
+                throw new CustomError('EarlyHangup',
+                  dialed);
+              }
+            });
+        });
+    })
+  .catch(function (err) {
+    if(err.message !== 'Inbound channel hungup') {
```
Comment (samuelg) [inline lib/sla.js]: You should use a custom error and check err.name instead. Error messages can change, error names should not.

**Comment by samuelg** [comment]
Looks good. Make sure to squash into a single commit before merging.
